### PR TITLE
Introduce procedure mode selection for RedisGraph.

### PR DESCRIPTION
**Pull request by haysch**
# Problem

It is not possible to use `GraphCache` for replicas-only connections.
If `GraphCache.GetCachedData` is invoked, the following exception will be thrown

```
StackExchange.Redis.RedisServerException: READONLY You can't write against a read only replica.
   at NRedisStack.Auxiliary.ExecuteAsync(IDatabaseAsync db, SerializedCommand command)
   at NRedisStack.GraphCommandsAsync.QueryAsync(String graphName, String query, Nullable`1 timeout)
   at NRedisStack.GraphCommandsAsync.CallProcedureAsync(String graphName, String procedure, IEnumerable`1 args, Dictionary`2 kwargs)
   at NRedisStack.GraphCommandsAsync.CallProcedureAsync(String graphName, String procedure)
```

# Context

Current implementation uses `Query` and `QueryAsync` for every `CALL PROCEDURE`.
However, as procedures can be defined as `write` or `read` (as defined by `dbms.procedures()`), it would make sense to be able to select mode.

# Solution

Add option to specify procedures as Read or Write when calling `CallProcedure` and `CallProcedureAsync`.

## Discussion

I opted to go for the enum approach, and specify `ProcedureMode.Write` as default to avoid breaking existing use.
One could instead go with a new `CallReadProcedure` and `CallReadProcedureAsync`.

### CODE REVIEW

**Comment by shacharPash** [comment]
Hi @haysch, thanks for your contribution!
I hope to go through the PR in the coming days.
In the meantime, I noticed that the coverage has dropped a bit, can you fix it?

**Comment by haysch** [comment]
Sure thing, will look into it as soon as possible.

Please let me know what you think of the procedure mode discriminator.
As stated I was a bit in doubt what would fit best - an enum or a new method.

File: src/NRedisStack/Graph/GraphCommandsAsync.cs [hunk 1]
```diff
@@ -56,22 +56,21 @@ public async Task<ResultSet> RO_QueryAsync(string graphName, string query, long?
...
-        await CallProcedureAsync(graphName, procedure, args, EmptyKwargsDictionary);
+        public Task<ResultSet> CallProcedureAsync(string graphName, string procedure, IEnumerable<string> args, ProcedureMode procedureMode = ProcedureMode.Write) =>
+            CallProcedureAsync(graphName, procedure, args, EmptyKwargsDictionary, procedureMode);
 
         /// <inheritdoc/>
-        public async Task<ResultSet> CallProcedureAsync(string graphName, string procedure, IEnumerable<string> args, Dictionary<string, List<string>> kwargs)
+        public async Task<ResultSet> CallProcedureAsync(string graphName, string procedure, IEnumerable<string> args, Dictionary<string, List<string>> kwargs, ProcedureMode procedureMode = ProcedureMode.Write)
         {
-            args = args.Select(a => QuoteString(a));
+            args = args.Select(QuoteString);
```
Comment (shacharPash) [inline src/NRedisStack/Graph/GraphCommandsAsync.cs:73]: As I understand this change doesn't really change the meaning of this line.
If you prefer this syntax, I think it should be the same as the line in GraphCommands.cs

File: src/NRedisStack/Graph/GraphCommands.cs [hunk 2]
```diff
@@ -55,20 +55,19 @@ public ResultSet RO_Query(string graphName, string query, long? timeout = null)
...
-        public ResultSet CallProcedure(string graphName, string procedure, IEnumerable<string> args) =>
-        CallProcedure(graphName, procedure, args, EmptyKwargsDictionary);
+        public ResultSet CallProcedure(string graphName, string procedure, IEnumerable<string> args, ProcedureMode procedureMode = ProcedureMode.Write) =>
+            CallProcedure(graphName, procedure, args, EmptyKwargsDictionary, procedureMode);
 
         /// <inheritdoc/>
-        public ResultSet CallProcedure(string graphName, string procedure, IEnumerable<string> args, Dictionary<string, List<string>> kwargs)
+        public ResultSet CallProcedure(string graphName, string procedure, IEnumerable<string> args, Dictionary<string, List<string>> kwargs, ProcedureMode procedureMode = ProcedureMode.Write)
         {
             args = args.Select(a => QuoteString(a));
```
Comment (shacharPash) [inline src/NRedisStack/Graph/GraphCommands.cs:72]: Change this to `args = args.Select(QuoteString);` OR cancel the change in GraphCommandsAsync.cs

File: src/NRedisStack/Graph/GraphCommandsAsync.cs [hunk 3]
```diff
@@ -56,22 +56,21 @@ public async Task<ResultSet> RO_QueryAsync(string graphName, string query, long?
...
 
-        internal static readonly Dictionary<string, List<string>> EmptyKwargsDictionary =
+        private static readonly Dictionary<string, List<string>> EmptyKwargsDictionary =
             new Dictionary<string, List<string>>();
 
-        // TODO: Check if this is needed:
         /// <inheritdoc/>
-        public async Task<ResultSet> CallProcedureAsync(string graphName, string procedure) =>
-        await CallProcedureAsync(graphName, procedure, Enumerable.Empty<string>(), EmptyKwargsDictionary);
+        public Task<ResultSet> CallProcedureAsync(string graphName, string procedure, ProcedureMode procedureMode = ProcedureMode.Write) =>
```
Comment (shacharPash) [inline src/NRedisStack/Graph/GraphCommandsAsync.cs:63]: Why did you delete the async await here?

File: src/NRedisStack/Graph/GraphCommandsAsync.cs [hunk 4]
```diff
@@ -56,22 +56,21 @@ public async Task<ResultSet> RO_QueryAsync(string graphName, string query, long?
...
+            CallProcedureAsync(graphName, procedure, Enumerable.Empty<string>(), EmptyKwargsDictionary, procedureMode);
 
         /// <inheritdoc/>
-        public async Task<ResultSet> CallProcedureAsync(string graphName, string procedure, IEnumerable<string> args) =>
-        await CallProcedureAsync(graphName, procedure, args, EmptyKwargsDictionary);
+        public Task<ResultSet> CallProcedureAsync(string graphName, string procedure, IEnumerable<string> args, ProcedureMode procedureMode = ProcedureMode.Write) =>
```
Comment (shacharPash) [inline src/NRedisStack/Graph/GraphCommandsAsync.cs:67]: async/await ?

File: src/NRedisStack/Graph/GraphCommandsAsync.cs [hunk 5]
```diff
@@ -9,7 +9,7 @@ namespace NRedisStack
 {
     public class GraphCommandsAsync : IGraphCommandsAsync
     {
-        IDatabaseAsync _db;
+        readonly IDatabaseAsync _db;
```
Comment (shacharPash) [inline src/NRedisStack/Graph/GraphCommandsAsync.cs:12]: What is the purpose of the change to readonly?

File: src/NRedisStack/Graph/GraphCommands.cs [hunk 6]
```diff
@@ -8,7 +8,7 @@ namespace NRedisStack
 {
     public class GraphCommands : GraphCommandsAsync, IGraphCommands
     {
-        IDatabase _db;
+        readonly IDatabase _db;
```
Comment (shacharPash) [inline src/NRedisStack/Graph/GraphCommands.cs:11]: readonly ?

File: tests/NRedisStack.Tests/Graph/GraphTests.cs [hunk 7]
```diff
@@ -956,6 +956,49 @@ public void TestModulePrefixs()
...
+    public void TestCallProcedureDbLabels()
+    {
+        var db = redisFixture.Redis.GetDatabase();
+        db.Execute("FLUSHALL");
+
+        const string graphName = "social";
+
+        var graph = db.GRAPH();
+        // Create empty graph, otherwise call procedure will throw exception
+        graph.Query(graphName, "RETURN 1");
```
Comment (shacharPash) [inline tests/NRedisStack.Tests/Graph/GraphTests.cs:969]: trailing space, It's worth checking the rest of the code as well and making sure there are no more trailing spaces.

**Review by shacharPash — CHANGES_REQUESTED** [review]
Hi, thanks again for your contribution, looks good!
I wrote some comments and questions about the code. I would appreciate it if you could answer my questions and change what is needed :)

**Review by haysch — COMMENTED** [review]
First batch of answers.
Will take a look at the remaining suggestions later :)

File: tests/NRedisStack.Tests/Graph/GraphTests.cs (hunk 7 above)
Comment (haysch) [reply tests/NRedisStack.Tests/Graph/GraphTests.cs:969]: I can find trailing whitespace in `src/NRedisStack/ResponseParser.cs`,  `tests/Doc/SearchQuickStartExample.cs` and `tests/Doc/SetGetExample.cs`.
Do you want me to trim those as well or create a separate PR that is more inline with #116?

**Comment by haysch** [comment]
Is it possible to retry the test run without pushing a dummy commit?
It seems like something had gone wrong in a completely unrelated test 🤔

File: tests/NRedisStack.Tests/Graph/GraphTests.cs (hunk 7 above)
Comment (shacharPash) [reply tests/NRedisStack.Tests/Graph/GraphTests.cs]: I think it's better separately to keep the PR clean.
But in the code of this PR there should be no trailing spaces.

File: src/NRedisStack/Graph/GraphCommands.cs (hunk 2 above)
Comment (haysch) [reply src/NRedisStack/Graph/GraphCommands.cs]: Changed to use method call.

File: src/NRedisStack/Graph/GraphCommands.cs [hunk 8]
```diff
@@ -81,7 +80,9 @@ public ResultSet CallProcedure(string graphName, string procedure, IEnumerable<s
                 queryBody.Append(string.Join(",", kwargsList));
```
[the remark this answers is not in the document]
Comment (shacharPash) [reply src/NRedisStack/Graph/GraphCommands.cs:80]: I also can't think of other options besides YEILD,
Maybe in the past there were more options and the code has remained that way ever since.
What would you like to change it to?

**Comment by shacharPash** [comment]
@slorello89 do you want to take a look?

File: src/NRedisStack/Graph/GraphCommands.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (haysch) [reply src/NRedisStack/Graph/GraphCommands.cs:80]: Even the lookup on the very specific key `y` seems tribal :)
If `YIELD` is the only possible option, it seems a bit much having to define a dictionary that contains a single key `y`.
So, in the spirit of the other arguments, a list of output parameters would make sense, and if it contains any values, then we add `YIELD` to the query.

Otherwise, if we were to keep the dictionary approach as-is, then the mapping could be `key => value`, where `key` is name of a clause, i.e., `YIELD`, and the `value` is a list of parameters, i.e., `val0, val1, val2.
However, I am not sure what other clauses there are available in a CALL procedure query.

[the remark this answers is not in the document]
Comment (shacharPash) [reply src/NRedisStack/Graph/GraphCommands.cs:80]: After consulting with @chayim, I think we'll leave it as it is for now.
thanks anyway!

**Review by shacharPash — APPROVED** [review]
LGTM 👍
